tokenscan/scanner.go

```diff
@@ -0,0 +1,221 @@
+package tokenscan
+
+import (
+	"context"
+	"math"
+	"math/rand"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/gocql/gocql"
+
+	"github.com/scylladb/gocqlx"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/scylladb/gocqlx/qb"
+)
+
+const (
+	// parallelism = nodesInCluster * coresInNode * 3
+	defaultParallelism = 72
+)
+
+type tokenRange struct {
+	startRange int64
+	endRange   int64
+}
+
+// SessionBuilder creates a gocql.Session
+type SessionBuilder interface {
+	CreateSession() (*gocql.Session, error)
+}
+
+// ApplyFunc is a callback function that is called with each retrieved scan item.
+// The scan is aborted if the ApplyFunc returns false
+type ApplyFunc func(context.Context, map[string]interface{}) bool
```

Comment:
I do not think that context should be passed here, implementers can use context of their own.
This implies usage of MapScan I'd like to see StructScan, therefore ApplyFunc should not exist and we should use interface{} like in cmp for instance.

```
// SortSlices returns a Transformer option that sorts all []V.
// The less function must be of the form "func(T, T) bool" which is used to
// sort any slice with element type V that is assignable to T.
//
// The less function must be:
//	• Deterministic: less(x, y) == less(x, y)
//	• Irreflexive: !less(x, x)
//	• Transitive: if !less(x, y) and !less(y, z), then !less(x, z)
//
// The less function does not have to be "total". That is, if !less(x, y) and
// !less(y, x) for two elements x and y, their relative order is maintained.
//
// SortSlices can be used in conjunction with EquateEmpty.
func SortSlices(lessFunc interface{}) cmp.Option {
	vf := reflect.ValueOf(lessFunc)
	if !function.IsType(vf.Type(), function.Less) || vf.IsNil() {
		panic(fmt.Sprintf("invalid less function: %T", lessFunc))
	}
	ss := sliceSorter{vf.Type().In(0), vf}
	return cmp.FilterValues(ss.filter, cmp.Transformer("cmpopts.SortSlices", ss.sort))
}
```